### PR TITLE
Javascript Runtime Reference

### PR DESCRIPTION
**Pull request by pytlesk4**
Re wrote it

Have fun reviewing this @rowa97 hahaha

### CODE REVIEW

**Comment by marbemac** [comment]
jsdoc

**Comment by marbemac** [comment]
big objects to bottom, link to them where needed. use h1-h4 headers for anything that should be in the sidebar toc that we will eventually have.

File: articles/prism/runtime.md [hunk 1]
```diff
@@ -1 +1,1503 @@
+# Introduction
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Should be updated to the title of the page. "Prism Javascript Runtime Reference"?

File: articles/prism/runtime.md [hunk 2]
```diff
@@ -1 +1,1503 @@
...
 
+Prism is really good at mocking and testing APIs. It is fast, flexible and easy to use. The scenarios and oas specifications have their limits, which is why Prism has a Javascript Runtime. Most of the time you won't need to jump into the runtime, and you should ask yourself before doing so if you have too use it. The runtime is powerful, and we chose javascript because of its flexibility and familiarity. It is perfect if you need to get/set/add/del(/eat?) cookies, loop over a http response body or seed your tests with some data.
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: This paragraph needs to be re-worked. Maybe something like what's below.

---

Prism is a fast, flexible, and easy-to-use application for mocking and testing HTTP API's. To help improve API testing effectiveness and bypass limitations in the OpenAPI specification, Prism comes with a full ES5-compatible Javascript runtime capable of supercharging your API workflow. 

Some example use cases for the Prism Javascript runtime include:

* Getting, setting, removing cookies on requests
* Manipulating HTTP request content, including slicing and dicing response bodies or manipulating headers
* Seeding your tests with randomized data that conforms to a regex

Anything that can be done in Javascript can now be done within a Scenario step.

File: articles/prism/runtime.md [hunk 3]
```diff
@@ -1 +1,1503 @@
...
+
+## Execution State
+
+The current state of prism dictates what is made available in the runtime. At the end of the day, Prism isn't really a proxy or a test runner. It is an orchestration engine, so when we are talking about state, we are talking about what is currently being orchestrated. Today Prism can orchestrate http, javascript and reffed scenarios.
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Prism should always be capitalized.

This:

> At the end of the day, Prism isn't really a proxy or a test runner.  It is an orchestration engine, so when we are talking about state, we are talking about what is currently being orchestrated. Today Prism can orchestrate http, javascript and reffed scenarios.

Should be updated to something like this:

> Prism is an orchestration engine as much as it is a proxy and test runner. When we talk about execution state, we are talking about what is currently being orchestrated. At this time, Prism can orchestrate HTTP, Javascript, other and referenced scenarios.

File: articles/prism/runtime.md [hunk 4]
```diff
@@ -1 +1,1503 @@
...
+
+### HTTP
+
+### Scripts
+
+### Refs
+
+# Javascript Specification
+
+EMacs 5/5.1
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Maybe turn this into a sentence:

> The Prism Javascript Runtime is compliant with the ECMAScript 5/5.1 specification, with full ECMAScript 6 support coming soon.

File: articles/prism/runtime.md [hunk 5]
```diff
@@ -1 +1,1503 @@
...
+
+# Globals
+
+#### Lodash
+
+[Lodash v4.17.4](https://lodash.com/) - It is always loaded in the runtime as `_`.
+
+#### Console
+
+The console objects gives you access to the Scenario Collection Logs. Use it to debug your sceanrios.
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Typo: sceanrios

File: articles/prism/runtime.md [hunk 6]
```diff
@@ -1 +1,1503 @@
...
+// TODO: Support default responses
+// TODO: Check to see if options is hooked up
+// TODO: If hooked up document it.
+// Returns HTTPOperation -> https://next.stoplight.io/stoplight/hubs-spec#/definitions/HTTP_Operation
+SL.specs.findOperation => function(path, method, options)
+```
+
+# Double Dollar - $$
+
+Double Dollar sign is the collection object.
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Don't capitalize Dollar here, unless we always want to refer to it as "Double Dollar Sign". We just need to be consistent.

File: articles/prism/runtime.md [hunk 7]
```diff
@@ -1 +1,1503 @@
...
+ *
+ * @param {string} string - The path to set. Will try to create appropriate objects if path doesn't exist.
+ * @param {Object} object - The value to set at the given path on your environment.
+ */
+$$.env.set(string, object);
+```
+
+#### Request
+
+The request object on double dollar sign is a special object to help with Prism Instances. Anytime a request is sent through prism and a prism instance is being served, that incoming request will be parsed and set on the double dollar sign request method. Double Dollar Request is defined in Prism Conductor Instances, but it won't have any affect on the Scenario Collection run.
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Capitalize prism.

File: articles/prism/runtime.md [hunk 8]
```diff
@@ -1 +1,1503 @@
...
+
+/**
+ * Del the current request's headers at key.
+ *
+ * @param {string} key - The key to delete on the headers string.
+ */
+$$.request.headers.del(key);
+```
+
+##### Content Types Content Negotiation
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Add a slash: Content Types / Content Negotiation

File: articles/prism/runtime.md [hunk 9]
```diff
@@ -1 +1,1503 @@
...
+```js
+/**
+ *
+ */
+tests => object;
+```
+
+## HTTP
+
+#### Step Request aka input
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: What does this mean?

File: articles/prism/runtime.md (hunk 6 above)
Comment (pytlesk4) [reply articles/prism/runtime.md]: Updating

File: articles/prism/runtime.md (hunk 7 above)
Comment (pytlesk4) [reply articles/prism/runtime.md]: Updating

File: articles/prism/runtime.md (hunk 9 above)
Comment (pytlesk4) [reply articles/prism/runtime.md]: HTTP Scenario Step type, aka input which is how you interact with it in the run time.

File: articles/prism/runtime.md (hunk 8 above)
Comment (pytlesk4) [reply articles/prism/runtime.md]: Updating

**Comment by pytlesk4** [comment]
ready for review

File: articles/prism/runtime.md [hunk 10]
```diff
@@ -1 +1,1554 @@
...
+
+* Getting, setting, removing cookies on requests
+* Manipulating HTTP request content, including slicing and dicing response bodies or manipulating headers
+* Seeding your tests with randomized data that conforms to a regex
+
+Anything that can be done in Javascript can now be done within a Scenario step.
+
+## Execution State
+
+Prism is an orchestration engine as much as it is a proxy and test runner. When we talk about execution state, we are talking about what is currently being orchestrated. At this time, Prism can orchestrate HTTP, Javascript, other and referenced scenarios.
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: > Prism can orchestrate HTTP, Javascript, other and referenced scenarios.

Consider updating to:

> Prism can orchestrate HTTP calls, Javascript programs, and other referenced scenarios.

File: articles/prism/runtime.md [hunk 11]
```diff
@@ -1 +1,1554 @@
...
+
+## Globals
+
+### Lodash
+
+[Lodash v4.17.4](https://lodash.com/) - It is always loaded in the runtime as `_`.
+
+### Console
+
+The console objects gives you access to the Scenario Collection Logs. Use it to debug your sceanrios.
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: sceanrios -> scenarios

File: articles/prism/runtime.md [hunk 12]
```diff
@@ -1 +1,1554 @@
...
+* @param {string} method - Reqeust method to search for in OAS connected Specfication.
+*
+* @returns {Operation} - Stoplight opeartion that matches given path and method.
+*/
+var operation = SL.specs.findOperation => function(path, method)
+```
+
+# Double Dollar - $$
+
+Double dollar sign is the collection object.
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: > Double dollar sign is the collection object.

Maybe update to:

> The double dollar sign variable represents the currently running collection object.

File: articles/prism/runtime.md [hunk 13]
```diff
@@ -1 +1,1554 @@
...
+ *
+ * @param {string} path - Path is the location of the value to set on the current request. If path doesn't exist then try to create it.
+ * @param {*} value - The value to set at the given path.
+ */
+$$.request.body.set(path, value);
+```
+
+#### Response
+
+Double dollar response is what is returned by Prism. There is a catch though, conducting scenarios outside of the Stoplight app if double dollar response isn't set, then the result of the scenario run is returned.
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Can you clarify the second sentence? I'm not sure what you're trying to say.

File: articles/prism/runtime.md [hunk 14]
```diff
@@ -1 +1,1554 @@
...
+/**
+ * Decodes a string of data which has been encoded using base-64 encoding.
+ *
+ * @param {string} string - the base-64 encoded string to decode.
+ * @returns {string} string - the original string.
+ */
+Base64.decode(string);
+```
+
+## Stoplight Object (SL)
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Should this be under the Global section?

File: articles/prism/runtime.md [hunk 15]
```diff
@@ -1 +1,1554 @@
...
+
+#### Sleep
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: We should be consistent with header depth, so maybe make this a `###` instead? This applies to the other 4 headers below.

File: articles/prism/runtime.md [hunk 16]
```diff
@@ -1 +1,1554 @@
...
+*
+* @param {string} path - Request URL Path to search for in OAS connected Specification.
+* @param {string} method - Reqeust method to search for in OAS connected Specfication.
+*
+* @returns {Operation} - Stoplight opeartion that matches given path and method.
+*/
+var operation = SL.specs.findOperation => function(path, method)
+```
+
+# Double Dollar - $$
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Should this be under the Globals section?

File: articles/prism/runtime.md [hunk 17]
```diff
@@ -1 +1,1554 @@
...
+
+/**
+ * Del the current request's headers at key.
+ *
+ * @param {string} key - The key to delete on the headers string.
+ */
+input.headers.del(key);
+```
+
+##### Content Types Content Negotiation
```
Comment (rossmcdonald) [inline articles/prism/runtime.md]: Maybe update to "Content Types / Content Negotiation"

**Pull request merged**